pyresample/geometry.py

```diff
@@ -717,37 +717,282 @@ def _do_transform(src, dst, lons, lats, alt):
-        res = res.coarsen(**dims).mean()
+                            src_lons.data,
+                            src_lats.data,
+                            da.zeros_like(src_lons),  # altitude
+                            new_axis=[2],
+                            chunks=(src_lons.chunks[0], src_lons.chunks[1], 3))
+        res = xr.DataArray(res, dims=['y', 'x', 'xyz'], coords=src_lons.coords)
+
+        # Aggregating
+        res = res.coarsen(x=x, y=y, **kwargs).mean()
+
+        # Back-conversion to geographic CRS
         lonlatalt = da.map_blocks(self._do_transform, geocent, latlong,
-                                  res[:, :, 0].data, res[:, :, 1].data,
-                                  res[:, :, 2].data, new_axis=[2],
+                                  res[:, :, 0].data,  # x
+                                  res[:, :, 1].data,  # y
+                                  res[:, :, 2].data,  # z
+                                  new_axis=[2],
                                   chunks=res.data.chunks)
-        lons = DataArray(lonlatalt[:, :, 0], dims=self.lons.dims,
-                         coords=res.coords, attrs=self.lons.attrs.copy())
-        lats = DataArray(lonlatalt[:, :, 1], dims=self.lons.dims,
-                         coords=res.coords, attrs=self.lons.attrs.copy())
-        try:
-            resolution = lons.attrs['resolution'] * ((dims.get('x', 1) + dims.get('y', 1)) / 2)
-            lons.attrs['resolution'] = resolution
-            lats.attrs['resolution'] = resolution
-        except KeyError:
-            pass
+
+        # Back-conversion array as input format
+        lons, _ = _convert_2D_array(lonlatalt[:, :, 0], to=src_lons_format, dims=['y', 'x'])
+        lats, _ = _convert_2D_array(lonlatalt[:, :, 1], to=src_lats_format, dims=['y', 'x'])
+
+        # Add additional info if the source array is a DataArray
+        if isinstance(self.lats, xr.DataArray) and isinstance(self.lons, xr.DataArray):
+            lats = lats.assign_coords(res.coords)
+            lats.attrs = self.lats.attrs.copy()
+            lons = lons.assign_coords(res.coords)
+            lons.attrs = self.lons.attrs.copy()
+            try:
+                resolution = lons.attrs['resolution'] * ((x + y) / 2)
+                lons.attrs['resolution'] = resolution
+                lats.attrs['resolution'] = resolution
+            except KeyError:
+                pass
+
+        # Return the downsampled swath definition
+        return SwathDefinition(lons, lats)
+
+    def upsample(self, x=1, y=1):
+        """Upsample the swath definition along x (along-track) and y (cross-track) dimensions.
+
+        To upsample of a factor of 2 (each pixel splitted in 4 pixels), call swath_def.upsample(x=2, y=2).
+        swath_def.upsample(x=1, y=1) simply returns the current swath_def.
+        """
+        # TODO: An alternative would be to use geotiepoints.geointerpolator.GeoInterpolator
```

Comment:
Indeed, this should be using geotiepoints to avoid functionality duplication.

Reply:
I wasn't able to use geotiepoints to do that. I would need help to do that. 
Look at the code snippet in the PR ... I get strange results. 
This implementation btw is dask-friendly compared to geotiepoints ... 
It can even made more efficient (see TODOs in the code)

Reply:
I will have a look and try to help asap. It could be that geotiepoints is not working optimally, so we should improve it to make it work here I think.

Reply:
I spent 3 hours yesterday again trying to use geotiepoints for this use case, but I failed. 
I think the code I added can be made dask-friendly very easily. Have a look at the TODOs  in the code.